src/org/thoughtcrime/securesms/qr/QrShowFragment.java

```diff
@@ -119,6 +129,31 @@ public void onReceive(Context context, Intent intent) {
         return view;
     }
 
+    public void shareQr() {
+        try {
+            File file = new File(getActivity().getCacheDir(), "share-qr.png");
+            file.createNewFile();
+            file.setReadable(true, false);
+            FileOutputStream stream = new FileOutputStream(file);
+            bitmap.compress(Bitmap.CompressFormat.PNG, 90, stream);
+            stream.flush();
+            stream.close();
+            Uri uri = FileProviderUtil.getUriFor(getActivity(), file);
+            Intent intent = new Intent(Intent.ACTION_SEND);
+            intent.setType("image/png");
+            intent.putExtra(Intent.EXTRA_STREAM, uri);
+            intent.putExtra(Intent.EXTRA_TEXT, hint);
+            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
+            startActivity(Intent.createChooser(intent, getString(R.string.chat_share_with_title)));
```

Comment:
I just found the method `DcHelper.openForViewOrShare()` which does the same thing but with more special-casing e.g. for older Android versions. (also, it takes a message id, not a file, but we could extract everything after the line `String mimeType = msg.getFilemime();` into an extra function that can then share an arbitrary file.

With luck, this would also mean that we don't have to change `file_provider_paths.xml` because I'm not sure what the security implications of this change are - I hope that it doesn't give other apps full access to our cache directory, but Android tends to surprise me (except if you know @adbenitez?). And also, there is probably a reason that we do the thing with `DcHelper.sharedFiles`.

Reply:
yip, using sth. as in openForViewOrShare() sounds like the best way to to that, thanks for pointing that out. sounds not to be that hard.

i agree with having mixed feelings about possible security implications of security of `file_provider_paths.xml` ...

as a quick step to move forward, we can also revert the commit https://github.com/deltachat/deltachat-android/pull/2103/commits/2a2a801b311e7c4929f5063d3340234413528b93 that introduces the issue - seems as if also with getExternalCacheDir things work in general.